### PR TITLE
[server] For GitLab projects without an owner avatar, fall back to the namespace avatar, or generate the default GitLab avatar

### PR DESCRIPTION
**Pull request by jankeromnes**
## Description

For GitLab projects belonging to a group, it seems that there is no `owner.avatar_url` to be found here:

https://github.com/gitpod-io/gitpod/blob/fc68e4780eae27c5a3e7c0907f457ebb393a0371/components/server/ee/src/gitlab/gitlab-app-support.ts#L46

Example group project from GitLab's API:

```ts
{
  "id": 23116976,
  "description": "",
  "name": "sample-project",
  "name_with_namespace": "reproduction-group / sample-project",
  /* ... (no "owner") ... */
  "namespace": {
    "id": 10408914,
    "name": "reproduction-group",
    /* ... */
    "avatar_url": "/uploads/-/system/group/avatar/10408914/gitpod-ua.png",
    "web_url": "https://gitlab.com/groups/reproduction-group"
  },
  /* ... */
}
```

As you can see, there is no `"owner"`, but there is an avatar URL on the `"namespace"` (although it's relative).

When there is no owner avatar URL, this Pull Request falls back to the namespace avatar URL if available.

EDIT: When no avatar is available, it generates a SVG avatar that looks like GitLab's own default group avatar.

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/5330

## How to test

1. Connect with GitLab
2. Be part of a group which has an avatar (e.g. https://gitlab.com/reproduction-group)
3. Go to `/new` -- the GitLab group should have the set avatar visible

## Release Notes

```release-note
[server] For GitLab projects without an owner avatar, fall back to the namespace avatar, or generate the default GitLab avatar
```

### CODE REVIEW

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (andrew-farries) [inline components/server/ee/src/gitlab/gitlab-app-support.ts:65]: Is it possible to do any better than an `any` type here?

Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:65]: Good catch!

The original code uses `const anyProject = project as any`, so I've just used that in this method.

It looks like the types returned by the API are of type `ProjectSchema`, but also that type seems incomplete:

Screenshot 2022-03-15 at 14 40 01

Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:65]: We could potentially reverse-engineer the type, e.g.:

```ts
type ProjectSchema = {
    owner?: {
        avatar_url: string;
    },
    namespace: {
        avatar_url: string;
    },
};
```

but I'm not sure if this is better than `any`. 🤔

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [inline components/server/ee/src/gitlab/gitlab-app-support.ts:70]: **issue**: Not sure if this is the correct line to comment this but the avatar seems broken now while before it was just a white blank space, see screenshots below.

| BEFORE |  AFTER|
|-|-|
| list-before | list-after |

**Review by gtsiolis — COMMENTED** [review]
Thanks for picking this up, @jankeromnes!

Commented below about two outstanding issues related to UX here.

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:66]: **thought**: Alternatively, we could introduce a custom fallback image using GitLab's logo for the groups that we cannot fetch their custom generated avatar, see [relevant comment](https://github.com/gitpod-io/gitpod/issues/5330#issuecomment-1067954451):

> GitLab auto-generates an avatar based on the group name and using a colored background in combination with the first letter of the group name.

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:70]: Ah, thanks for catching this @gtsiolis. If you inspect the broken icons, what does the avatar URL look like?

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:66]: Not known. Would you know how I can become part of such a group with less permissions, in order to reproduce this?

Alternatively, I can add debug logs to this PR, and ask you to load that page again. 😊

Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:66]: Also, I agree that the fallback image would be nice, but [I couldn't find how to generate it](https://github.com/gitpod-io/gitpod/pull/8824#issuecomment-1067956035):

> Note: When there is no avatar, it would be nice to use the same default avatar as GitLab (i.e. a big letter with a pastel background). However, I couldn't find how to generate this, so for now I left the avatar URL empty when there is no avatar.

Would you know a potential way to find out? 👀

Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:66]: @jankeromnes Last time I remember contributing to this[[1](https://gitlab.com/gitlab-org/gitlab-foss/-/merge_requests/17271)], GitLab was using a custom logic to create the icon, see [relevant code](https://gitlab.com/gitlab-org/gitlab/-/blob/b2a22b6e85200ce55ab09b5c765043441b086c96/app/helpers/avatars_helper.rb#L151-161). Maybe this is why this is not available through their API?

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (andrew-farries) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:65]: One option would be to get the data returned by the API and then use `typeof` to generate a type from it, if there is no 'official' specification available.

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts]: Ah, thank you so much @gtsiolis for linking the relevant code ❤️ I couldn't find it, and now I've used it to implement a similar defafult-avatar-generator for our GitLab avatar fetching code.

**Comment by jankeromnes** [comment]
Ready for another round of reviews! 🚀

I've implemented a default avatar generator similar to GitLab's one (thanks @gtsiolis for linking the relevant code! 🎯) so there should always be an avatar now.

I've also implemented more specific types (instead of using `any`), based on the official types from our library (`gitbeaker`) with the additional missing fields.

**Comment by gtsiolis** [comment]
Taking another look now! 👀

**Review by gtsiolis — COMMENTED** [review]
Took another look and left some more comments below. Thanks, @jankeromnes! 🏀

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts]: **praise**: Thanks for fixing this, work like a charm. ✨

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [inline components/server/ee/src/gitlab/gitlab-app-support.ts:92]: **issue**: One minor issue here is that some avatars are using a different background color from what GitLab does for each group. I cound not reproduce this for new groups but some of the existing groups. Is this expected?

| on GitLab.com | on Gitpod.io |
|-|-|
| Screenshot 2022-03-28 at 10 12 54 PM | Screenshot 2022-03-28 at 10 13 06 PM |

Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:92]: Interesting! Would have to look at debug logs -- maybe we're not using the right ID for these groups 👀

Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:92]: Yeah, probably also minor issue that's not worth blocking us from merging this. 💭

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:95]: They are supposed to be centered. I'll debug the produced SVG. 🔍

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:89]: Good point. I guess we could somehow re-use the same fonts as in our dashboard, but:
- Since they're not exactly the same as GitLab, I'm worries that in some cases they might look different than GitLab icons (which fails to achieve our goal here)
- Also, I'm not entirely sure how to reference the Tailwind class from inside the SVG (which doesn't use Tailwind) 🤔

So, maybe copying the fonts from GitLab just for the GitLab default avatar makes sense? 😇

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:88]: Not sure what the difference is. I'll try to find equivalent sources from the `gitlab` repo.

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:89]: To clarify, the question aimed to suggest not importing any fonts but relying on our font stack to render the character inside the group avatar. We do use the same background color which already helps, right?

Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:89]: This is what the SVG looks like for me when I remove the font specification:

Screenshot 2022-03-30 at 15 49 22

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:88]: > Not sure what the difference is.
The difference is that gitlab-foss does not contain any enterprise or licensed code, mostly everything inside the `ee/` top folder which is automatically dropped. Some open source orgs have a requirement of not using enterprise or licenced code and strictly use GitLab's FOSS version.

I've already linked the equivalent source on the `gitlab` repository in the code suggestion in https://github.com/gitpod-io/gitpod/pull/8824#discussion_r836723888.

Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:88]: Haha, since GitHub was being slow, it took me 5 minutes to understand that you actually linked to the comment right above mine 😅 fair enough, many thanks for finding the correct source!

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:89]: Oh, I meant rely on (use) our font stack instead. 😇

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:18]: Good catch! Found [a way to support emojis](https://dev.to/acanimal/how-to-slice-or-get-symbols-from-a-unicode-string-with-emojis-in-javascript-lets-learn-how-javascript-represent-strings-h3a) as first character. 👍

Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:18]: @gtsiolis Could you please try again, or invite me to that group? (For some reason, I'm not able to create one that starts with an emoji)

Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:18]: @jankeromnes It almost worked! I added this to the reproduction-group in GitLab (see [example project](https://gitlab.com/reproduction-group/sample-group/sample-project)), but somehow the emoji is surfacing to the top-level group. Previously, refrained from adding this as it would break new project wizard for you, too. 🤷  

Screenshot 2022-03-30 at 7 01 12 PM

**Comment by jankeromnes** [comment]
Briefly taking this back to Draft for some quick debugging.

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:19]: @gtsiolis Ah, I think I got it:
- Previously, we picked the avatar only for the `owner` (when provided by GitLab)
- With this Pull Request, we also look at the `namespace` when the `owner` is not available
- However, I see that `namespace` is not always the "root namespace" (i.e. for `reproduction-group / 🍊-sample-group / 🍋-sample-project`, GitLab provides `🍊-sample-group`, not `reproduction-group`)

I believe the fix is to only use the avatar for `owner` if provided, or `namespace` if it is the root namespace. (Note: We could also look at the `parent_id` of nested groups, then make follow-up GitLab API requests to find the root namespace and its avatar -- but I think that would be way over-engineered at this point.)

Let's see how the new algorithm behaves. Pushing soon.

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (jankeromnes) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:93]: I think this might also be a case of "using the avatar of a nested group": https://github.com/gitpod-io/gitpod/pull/8824#discussion_r839377129

**Comment by jankeromnes** [comment]
_Phew._ Okay now this should be ready to rock'n'roll. 🎸😁

@gtsiolis please take another look when you have some time. 🙏

(For what it's worth, I initially estimated this work as a 0.25 day effort / 1 iteration -- looks like I was about 10x off on both. 😅 )

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (easyCZ) [inline components/server/ee/src/gitlab/gitlab-app-support.ts:100]:
```suggestion
        const backgroundColors = ["#fcf1ef", "#f4f0ff", "#f1f1ff", "#e9f3fc", "#ecf4ee", "#fdf1dd", "#f0f0f0"];
        const backgroundColor = backgroundColors[owner.id % backgroundColors.length]
```
To avoid the future fix-up when a new color is added to the mix.

File: components/dashboard/src/projects/NewProject.tsx
Comment (easyCZ) [inline components/dashboard/src/projects/NewProject.tsx:240]: 👍 Much nicer on the eyes

**Review by easyCZ — APPROVED** [review]
Code-wise looks good to me. Adding hold to resolve remaining questions.

/hold

**Comment by gtsiolis** [comment]
Taking another look at this now! 👀

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts]: Looks resolved now. Thanks for fixing this, @jankeromnes!

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [inline components/server/ee/src/gitlab/gitlab-app-support.ts:102]: **praise**: Lovely comment! 💯

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:100]: This seems to be also resolved now!

**Review by gtsiolis — APPROVED** [review]
LGTM, @jankeromnes! 🔮

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts]: The groups with custom avatars still seems broken. What do you think of opening a follow up issue for this to keep track of this?

Frame 370

**Comment by jankeromnes** [comment]
Nits addressed & follow-ups filed! 🎉

@easyCZ or @gtsiolis could you please re-approve?

/unhold

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts]: It seems to be working ok now for all but one of the groups in there, posted a comment in https://github.com/gitpod-io/gitpod/issues/9112#issuecomment-1087787341.

**Review by gtsiolis — APPROVED** [review]
Works like a charm, @jankeromnes! 🔮 

Thanks for also opening the follow up issue in https://github.com/gitpod-io/gitpod/pull/8824#discussion_r841945398 to track one potentially minor edge case. 💯 

Approving for UX, but you'll still need an approval from someone from the WebApp team.
> Waiting on code owner review from gitpod-io/engineering-webapp.

**Review by andrew-farries — APPROVED** [review]

**Pull request merged**
